### PR TITLE
Farming: Trees: Convert saplings and plants to NodeTimerRef.

### PR DESCRIPTION
**Pull request by sofar**
This PR requires @minetest/minetest#3677

Farming and plant growth has traditionally in minetest been
implemented using ABM's. These ABM's periodically tick and cause
plants to grow. The way these ABM's work has several side effects
that can be considered harmful.

Not to mention a comprehensive list of downsides here, but ABM's
are chance-dependent. That results in the chance that some nodes
potentially never get processed by the ABM action, and others get
processed always. One can easily find this effect by planting a large
field of crops, and seeing that some nodes are fully grown really
fast, and some just won't make it to fully grown status even after
hours or play time.

One could solve the problem by making the ABM's slower, and giving them
a 100% of action, but this would cause the entire field to grow a step
instantly at ABM intervals, and is both ugly, and a large number of
node updates that needs to be sent out to each client. Very un-ideal.

With NodeTimers though, each node will see a separate node timer event,
and they will likely not coalesce. This means that we can stop relying
on chance to distribute plant growth, and assign a single timer event
to grow the plant to the next phase.  Due to the timer implementation,
we won't ever miss a growth event, and we can re-scehdule them until
the plant has reached full size.

Previously, plants would attempt to grow every 9 seconds, with a
chance of 1/20. This means typically, a plant would need 9_20 seconds
to grow 1 phase, and since there are 8 steps, a typical plant growth
would require 9_20_8 ABM node events. (spread out over 9_8 ABM actual
underlying events per block, roughly).

because plants are likely not growing to full for a very long time
due to statistics working against it (5% of the crops take 20x longer
than the median to grow to full, we'd be seeing ABMs fire possibly
up to 9_20_8*20 with a 95% confidence interval (the actual math
is likely off, but the scale should be correct). That's incredibly
wasteful. We'd reach those conditions easily with 20 plant nodes.

Now, after we convert to NodeTimers, each plant node will see exactly
8 NodeTimer events, and no more. This scales lineairly per plant.

An additional problem in the farming mod was that the final fully-grown
plant was also included in the ABM, causing infinite more ABM's even
after the entire field had grown to completion.

Now, we're left with the problem that none of the pre-existing plants
have actual node timers started on them, and we do not want a new ABM
to fix this issue, since that would be wasteful.  Fortunately, there
is now an LBM concept, and we can use it to assure that NodeTimers
on crop nodes are properly started, and only have to do the actual
conversion once per block, ever.

We want to provide a fairly similar growth rate after this conversion
and as such I've resorted to modelling some statistical data. For this
I created a virtual 32x32 crop field with 9 steps (8 transitions)
as is the default wheat crop. We then apply a step where 1 in 20
plants in the field grows a step (randomly chosen) and count the
number of steps needed to get to 25%, 50, 75% and 95% grown.

The resulting data looks as follows:

25% - ~120 steps \* 9 sec / abm = 1080s
50% - ~152 steps               = 1368s
75% - ~194 steps               = 1746s
95% - ~255 steps               = 2295s

Next, we want to create a model where the chance that a crop grows
is 100% every node timer. Since there will only be 8 steps ever,
we want the slowest crops to grow in intervals of ~ 2300 / 8 seconds
and the fastest 1/4 of crops to grow 1080 / 8 seconds intervals.
We can roughly compare this to a normal distribution with a median
of 1400 with a stddev of ~350 (thick fingering this one here).

The rest is a bit of thick-fingering to get similar growth rates,
taking into account that ABM's fire regularly so if they're missed
it's fairly painless, but our timers are going to be 1-2 minutes
apart at minimum. I calculate the timer should be around 150s
median, and experimented with several jitter ranges.

Eventually I settled for now on [80,200] with a redo of [40,80],
meaning that each growth step at minimum takes (80 to 200) seconds,
and if a negative growth condition was found (darkness, soil not
wet, etc), then the growth step is retried every (40 to 80) seconds.

The end result is a growth period from seed to full in ~ 2.25
minetest days. This is a little bit shorter than the current
growth rate but the chances you'll miss timer ticks is a bit
larger, so in normal gameplay it should be fairly comparable.

A side effect is that fields grow to full yield fairly quickly
after crops make it to mature growth, and no crops are mature
a very long time before the majority grows to full. The spread
and view over a growing field is also fairly even, there's no
large updates with plenty of nodes. Just a node here or there
every second or so in large fields.

Ultimately, we get rid of ABM rollercoasters that cause tens of
node updates every 9 seconds. This will help multiplayer servers
likely a lot.

### CODE REVIEW

**Comment by sofar** [comment]
@paramat I'm assuming you'd like to test this, as well as @est31.

**Comment by 0-afflatus** [comment]
I like the sound of this, my one caveat it that I like the randomness of growth time.

**Comment by sofar** [comment]
It's still entirely random, except there's no more crops mature in 160 seconds, and there's no more crops that *never* mature. The variation can also be increased ... values I used are just a starting point.

**Comment by tenplus1** [comment]
A new way to grow crops which may be a good way to go...  Have you checked the routines in Farming Redo ? they have a random growth also and can even grow when player is not around.

**Comment by paramat** [comment]
Very interesting and the removal of ABM code will be welcome.

**Comment by sofar** [comment]
@tenplus1 the randomness isn't a new aspect. 

Node Timers always catch up when they're expired, so if players get back to their fields after a long time, it's guaranteed to grow 1 step.

File: mods/farming/api.lua [hunk 1]
```diff
@@ -150,30 +158,87 @@ farming.place_seed = function(itemstack, placer, pointed_thing, plantname)
...
+	-- check if on wet soil
+	local below = minetest.get_node({x = pos.x, y = pos.y - 1, z = pos.z})
+	if minetest.get_item_group(below.name, "soil") < 3 then
+		tick_again(pos)
+		return
+	end
+
+	-- check light
+	local light = minetest.get_node_light(pos)
+	if not light or light < def.minlight or light > def.maxlight then
```
Comment (kilbith) [inline mods/farming/api.lua]:
```Lua
if not light or light ~= def.minlight then
```

**Comment by 0-afflatus** [comment]
This could be the WTG for a few node changes and we could cut out some unnecessary overheads.

**Comment by C1ffisme** [comment]
:+1: :+1: :+1: :+1:

This is an extremely good idea, just as it is with furnaces. The removal of ABMs will be greatly welcomed. Maybe one day servers might not even lag...

**Comment by tenplus1** [comment]
Could you change the grow_plant function from local to global (farming.grow_plant) so that mod makers can create their own crops without having to use the api and still have them grow ?!?!

**Comment by sofar** [comment]
@tenplus1 agreed, that's a good idea.

**Comment by sofar** [comment]
the grow functions is now global: ```farming.grow_plant(pos)```

**Comment by paramat** [comment]
Not sure this could be ready and stable for 0.4.14 but maybe.

**Comment by sofar** [comment]
Not without the LBM change merged, anyway.

**Comment by est31** [comment]
I plan to get the LBM patch merged until 0.4.14, the PR is looking well already.

**Comment by sofar** [comment]
@paramat this patch itself is relatively safe... Tweaking the NodeTimer rates for plants growing may be the only thing left, and can be done incrementally.

**Comment by sofar** [comment]
Tuned plants to grow to mature in just under ~3 days of gameplay (1hr). Slightly longer than before, but with the more consistent growth, I feel this is more balanced.

**Comment by C1ffisme** [comment]
@sofar 1 hour seems a bit slow, unless you have a 150*150 farm. I suppose it would be stupid to make plants grow quicker if there are less of them, but it would make a bit more gameplay balance.

**Comment by sofar** [comment]
Many plants are going to be done at the end of day 2, so it's not as bad as it sounds.

The 3-day grow length means that 95%+ will be done. That's faster than the current growth, where often plants take 4+ days to make and sometimes longer, due to the underlying algorithm.

A 150^2 farm is to large to grow nicely, ABMs and node timers won't work further than 32 nodes away, so not even 1/4 of that would actually grow if you'd sit in it.

**Comment by Gael-de-Sailly** [comment]
:+1:

**Comment by paramat** [comment]
3 days is good.

**Comment by C1ffisme** [comment]
@sofar Okay, I see your points.

**Comment by Fixer-007** [comment]
:+1:

**Comment by sofar** [comment]
I've added a second conversion to nodetimers:

```
Convert saplings to node timer.
    
    Each sapling is given a single node timer that is between
    2 and 4 days of game play time (40-80 minutes). If you walk out
    of the zone, and come back later, the tree will always grow
    to full if the timer has elapsed.
    
    Because trees.lua is all functions, it needs to be parsed before
    nodes.lua, since that references some of its functions. Hence,
    change the order of parsing here. Otherwise saplings would not
    grow to full.
```

**Comment by sofar** [comment]
Should have added, this uses an LBM as well to start the node timer for each sapling.

**Comment by sofar** [comment]
Can remove the needs-engine-change flag now!

File: mods/default/trees.lua [hunk 2]
```diff
@@ -59,57 +59,62 @@ end
...
 	end
-})
+end
 
+minetest.register_lbm({
+	name = "default:convert_saplings_to_node_timer",
+	nodenames = {"default:sapling", "default:junglesapling",
+			"default:pine_sapling", "default:acacia_sapling",
+			"default:aspen_sapling"},
+	action = function(pos, _)
```
Comment (kilbith) [inline mods/default/trees.lua]:
``` Lua
action = function(pos)
```

**Comment by kilbith** [comment]
Rigorous benchmarks (timers vs. ABM) : https://forum.minetest.net/viewtopic.php?p=178943#p178943

> Analysis/Conclusion
> 
> The timer-based solution is about 50% slower than the ABM-based solution while plants are growing (in >sufficient light) and while the player is present. [...]

**Comment by sofar** [comment]
Rebased. Two minor cleanups.

File: mods/default/trees.lua (hunk 2 above)
Comment (sofar) [inline mods/default/trees.lua]: check, got that.

**Comment by Ekdohibs** [comment]
:+1:
@kilbith While it may be a bit slower in some cases for now, I want to change the handling of node timers in the engine (using a priority queue instead of a list that is processed each step), that would make it a lot faster.

**Comment by sofar** [comment]
The ```Needs engine-change``` label isn't needed. This works well and already more efficient as the current ABM's keep firing even for fully grown plants. The optimization is nice, but unnecessary.

**Comment by yyt16384** [comment]
Is the [2400,4800] time interval for saplings too high, considering the old ABM runs every 500 seconds on average?

**Comment by sofar** [comment]
> Is the [2400,4800] time interval for saplings too high, considering the old ABM runs every 500 seconds on average?

No. I explain the math in detail in the commit message. ABM's may fire faster, but have a ```chance``` factor.

**Comment by paramat** [comment]
Removing from milestone in case many problems arise.

**Comment by asl97** [comment]
@sofar If you compare worse case (ie: a single sapling) then maybe it's better but what about best case?
If I place a whole stack of sapling (99) or more, chances are that one or two would grow and mining one of them take more than 10 seconds.
During the time it take me to mine one tree, more have would grown.

**Comment by sofar** [comment]
@asl97 The problem is that "best case" is statistically impossible due to poisson distribution curves. The "normal" case *is* the best case for ABMs. You can get lucky statistically only so much.

And the normal chances for ABMs are far worse than for nodetimers.

**Comment by sofar** [comment]
Rebased. Two minor cleanups to code style and lbm naming (shortened)

**Comment by asl97** [comment]
@sofar I may not know anything about poisson distribution curves but back when I was actively playing on a server, it was pretty normal for players to place at least 2 stacks of sapling and use a chainsaw to cut them down (with pipeworks gathering dropped item into a bunch of chest).

198+* tries with 1/50 chance, the odds doesn't seem that bad.

*it was a simple continuous collect sapling from chest, place all sapling and finally chainsaw trees and collect more sapling workflow, sometimes the number of sapling on the ground is more than 198.

**Comment by yyt16384** [comment]
If I'm doing the calculation correctly, with ABMs saplings have more than 99% chance to grow before 2400 seconds, while with node timers none will grow before that. (assuming block is always loaded of course)

**Comment by asl97** [comment]
what @yyt16384 said is correct, here is the [Binomial Distribution Formula](http://formulas.tutorvista.com/math/binomial-distribution-formula.html) written in python.

you can play around with the `n` value.
```python
import math
n=2400/10 # number of time the ABM would run in 2400 seconds
r=0 # the number of successful grow
c = (math.factorial(n)/(math.factorial(n-r)*math.factorial(r)))*((1/50)**r)*((49/50)**(n-r))
print("the chance of it not growing: %f%%"%(c*100))
print("the chance of it growing: %f%%"%((1-c)*100))
```
the chance of it not growing: 0.783895%
the chance of it growing: 99.216105%

____
```python
import math
n=198 # number of sapling
r=0 # the number of successful grow
c = (math.factorial(n)/(math.factorial(n-r)*math.factorial(r)))*((1/50)**r)*((49/50)**(n-r))
print("the chance of not even a single sapling growing: %f%%"%(c*100))
print("the chance of a single sapling growing: %f%%"%((1-c)*100))
````
the chance of not even a single sapling growing: 1.831315%
the chance of at least a single sapling growing: 98.168685% (also known as chance of a single best case)

**Comment by sofar** [comment]
ABM's are still less efficient, by a huge margin.

If you're complaining about the slower sapling growth time, then that's valid, but efficiency wise, ABM's are a waste of CPU cycles :)

**Comment by asl97** [comment]
> ABM's are still less efficient, by a huge margin.

@sofar That I am sure everyone can agree, especially when including most of the recent and pending improvement to the timer system.
____
> If you're complaining about the slower sapling growth time, then that's valid.

Three person including [myself](https://github.com/minetest/minetest_game/pull/870#issuecomment-219196789) has raise this issue.
[C1ffisme](https://github.com/minetest/minetest_game/pull/870#issuecomment-192078785), [yyt16384](https://github.com/minetest/minetest_game/pull/870#issuecomment-202067707)
____
How about putting minimum time needed to 50% probability and max at 80% probability?
(Although I am sure most would welcome a lower number)

type | minimum time | number of run | 50% probability | maximum time | number of run | 80% probabilty
----- | ----------------- | ------------------ | ------------------- | ------------------ | ----------------- | -------------------
plant | 126              | 14                    | 51.232502%    | 288                  | 32                  | 80.628852%
tree | 350               | 35                    | 50.692538%    | 800                  | 80                  | 80.135115%

in other words, for plant: 126, 288, for tree: 350, 800.
^ those are the smallest possible value above each of their probability base on their abm interval and chance.

**Comment by sofar** [comment]
I'm not changing the growth speed at this time. I think the slower speed is needed to make farming a bit more challenging and a welcome variation to an already too easy vanilla game.

Farming is "free stuff", and that should come at a cost. It was way too easy. You could plant 10 saplings and the first one would be fully grown by the time you planted the last one, that's too easy.

**Comment by yyt16384** [comment]
I have run some data before, and if you only consider "harvest every x seconds" type of strategy, the maximum efficiency you can get for plants is very close. So there are probably no problems with plants.

I suggest keeping a similar growth speed in this PR, and changing it in another PR so it can have more discussion.

**Comment by paramat** [comment]
> changing it in another PR so it can have more discussion.

Seems a good idea, then we'll have more feedback on a suitable speed.

**Comment by sofar** [comment]
I'm entirely unsure what the last 2 comments mean. Someone please explain if it means I should change this PR or if the modification of time should be separated.

note: this patch *has* to change the timing since the way time is dealt with mechanically is completely different, so this patch *has* to set new values.

**Comment by MT-Modder** [comment]
I think they mean merge this as is and then afterward tune speed.

**Comment by paramat** [comment]
> I think they mean merge this as is and then afterward tune speed.

Yes this.
Merged it will be tested more, so more feedback on suitable growing times.

**Comment by sofar** [comment]
Ah, thanks for clarifying that!

**Comment by paramat** [comment]
I intend to review this soon, hopefully it can be merged soon.

**Comment by paramat** [comment]
I think the author of farming should review the first commit.
Prize for most awesome commit message ever.

File: mods/farming/api.lua [hunk 3]
```diff
@@ -142,38 +151,94 @@ farming.place_seed = function(itemstack, placer, pointed_thing, plantname)
...
+	if not def.next_plant then
+		-- disable timer for fully grown plant
+		return
+	end
+
+	-- grow seed
+	if minetest.get_item_group(node.name, "seed") and def.fertility then
+		local soil_node = minetest.get_node_or_nil({x = pos.x, y = pos.y - 1, z = pos.z})
+		if not soil_node then
+			tick_again(pos)
```
Comment (paramat) [inline mods/farming/api.lua]: If there's no node below i don't think we should 'tick again' as this is unlikely to be fixed a little while later. This will reduce load.

File: mods/farming/api.lua [hunk 4]
```diff
@@ -142,38 +151,94 @@ farming.place_seed = function(itemstack, placer, pointed_thing, plantname)
...
+			if minetest.get_item_group(soil_node.name, v) ~= 0 then
+				minetest.swap_node(pos, {name = def.next_plant})
+				if minetest.registered_nodes[def.next_plant].next_plant then
+					tick(pos)
+					return
+				end
+			end
+		end
+
+		tick_again(pos)
```
Comment (paramat) [inline mods/farming/api.lua]: What is this 'tick again' needed for?

**Comment by paramat** [comment]
I have a vague sense commit 1 is good, it's complex though.

File: mods/farming/api.lua (hunk 3 above)
Comment (asl97) [inline mods/farming/api.lua]: Probably meant for an edge case, something like unloaded node like `ignore`, the plant has the `attached_node` group.

Comment (sofar) [inline mods/farming/api.lua]: Yes, I've been fairly conservative to "prune" dead plants that won't be able to grow, this one can for sure be dropped (you can't hoe from the side or top if there's a plant on dirt - it has to be under air)

File: mods/farming/api.lua (hunk 4 above)
Comment (sofar) [inline mods/farming/api.lua]: again, too conservative attitude to "try again later". I'll remove that one as well.

**Comment by sofar** [comment]
Updated commit one of two with addressed comments.

**Comment by paramat** [comment]
Ok i'm happy now :+1: for both commits, will merge soon.

File: mods/farming/api.lua [hunk 5]
```diff
@@ -142,38 +151,92 @@ farming.place_seed = function(itemstack, placer, pointed_thing, plantname)
...
+
+	if not def.next_plant then
+		-- disable timer for fully grown plant
+		return
+	end
+
+	-- grow seed
+	if minetest.get_item_group(node.name, "seed") and def.fertility then
+		local soil_node = minetest.get_node_or_nil({x = pos.x, y = pos.y - 1, z = pos.z})
+		if not soil_node then
```
Comment (asl97) [inline mods/farming/api.lua]: I really expect people to catch stuff like this,  this isn't checking for soil, `not minetest.get_node_or_nil` would only be true if the node isn't loaded, as in an `ignore` node.

this check probably shouldn't be here or needed since seed can only be place on soil as per the ` farming.place_seed` function.

File: mods/farming/api.lua [hunk 6]
```diff
@@ -142,38 +151,92 @@ farming.place_seed = function(itemstack, placer, pointed_thing, plantname)
...
+	end
+
+	-- grow
+	minetest.swap_node(pos, {name = def.next_plant})
+
+	-- new timer needed?
+	if minetest.registered_nodes[def.next_plant].next_plant then
+		tick(pos)
+	end
+	return
```
Comment (asl97) [inline mods/farming/api.lua]: unless I am wrong, the `return` here isn't needed.

File: mods/farming/api.lua (hunk 5 above)
Comment (sofar) [inline mods/farming/api.lua]: soil can turn to dirt, and it's entirely possible that there are mods that turn soil into dirt under various circumstances.

File: mods/farming/api.lua (hunk 6 above)
Comment (sofar) [inline mods/farming/api.lua]: it's written explicitly here to signal that no further timer events will happen.

File: mods/farming/api.lua (hunk 5 above)
Comment (sofar) [inline mods/farming/api.lua]: You're correct about the return value being wrong here, it indeed should tick again here.

**Comment by sofar** [comment]
Fixed the first of @asl97 comments: indeed wrongly omitted `tick_again(pos)` there.

**Comment by paramat** [comment]
Good point asl97.
:+1:

**Comment by paramat** [comment]
> I really expect people to catch stuff like this, this isn't checking for soil, not minetest.get_node_or_nil would only be true if the node isn't loaded, as in an ignore node. It shouldn't kill the timer just cause the node isn't loaded.

So even better how about: if 'get node or nil' is nil (unloaded node) then 'tick again', else if the node is not soil (node present but not soil) 'return'?

**Comment by yyt16384** [comment]
Note that it is possible with some mods (like mesecons pistons) to change the node below from non-soil to soil.

File: mods/farming/api.lua (hunk 5 above)
Comment (asl97) [inline mods/farming/api.lua]: you still don't get what I am talking about.  There is no such check for soil group in those two lines.
```
		local soil_node = minetest.get_node_or_nil({x = pos.x, y = pos.y - 1, z = pos.z})
		if not soil_node then
```
Where do you even see the check for the soil group?  Currently it is just checking whether the node is loaded or not loaded.
`minetest.get_node_or_nil` DOESN'T check for soil group.

**Comment by sofar** [comment]
@paramat that's what the current patch does.

@yyt16384 I'm leaving it explicitly in the current game that the plant stops growing if a piston is used to move soil back under the plant so that there's a way for people to craft non-growing crops for ornamental purposes.

File: mods/farming/api.lua [hunk 7]
```diff
@@ -142,38 +151,93 @@ farming.place_seed = function(itemstack, placer, pointed_thing, plantname)
...
+	-- grow seed
+	if minetest.get_item_group(node.name, "seed") and def.fertility then
+		local soil_node = minetest.get_node_or_nil({x = pos.x, y = pos.y - 1, z = pos.z})
+		if not soil_node then
+			tick_again(pos)
+			return
+		end
+		-- omitted is a check for light, we assume seeds can germinate in the dark.
+		for _, v in pairs(def.fertility) do
+			if minetest.get_item_group(soil_node.name, v) ~= 0 then
```
Comment (asl97) [inline mods/farming/api.lua]: @sofar 
> soil can turn to dirt, and it's entirely possible that there are mods that turn soil into dirt under various circumstances

.If you didn't want seed to "germinate" on dirt, you should probably check for `minetest.get_item_group(soil_node.name, v) >= 2`.
dirt has the soil group.

Comment (sofar) [inline mods/farming/api.lua]: The API allows this explicitly. There is no use for it now, but it would entirely be possible to plant e.g. flower seeds that grow to full on dirt.

I'm tempted to leave this as is.

**Comment by paramat** [comment]
> that's what the current patch does.

Oh yeah :]
:+1:

File: mods/farming/api.lua (hunk 7 above)
Comment (asl97) [inline mods/farming/api.lua]: oh wait.... soil **CAN'T** turn back to dirt if there is a plant or seed on it.
https://github.com/sofar/minetest_game/blob/ef3e5a8f6ccdfe37e8af02d29fea720d169a3558/mods/farming/nodes.lua#L135

**Comment by paramat** [comment]
70ef7864c1533cc50ef1da95985ff6b741778066
e0cb3fce02d172e477ed8bee3735fc7a13ed58c4

File: mods/farming/api.lua (hunk 7 above)
Comment (sofar) [inline mods/farming/api.lua]: >oh wait.... soil **CAN'T** turn back to dirt if there is a plant or
>seed on it.
>https://github.com/sofar/minetest_game/blob/ef3e5a8f6ccdfe37e8af02d29fea720d169a3558/mods/farming/nodes.lua#L135

Yes, but mods could add this behavior.

Comment (asl97) [inline mods/farming/api.lua]: @sofar the problem with replying through email is that you don't see the edits people might have made.
I added `If another mod does it, it would be their fault.` shortly after posting it.

**Comment by yyt16384** [comment]
Is it intended that timers are started in on_place (farming.place_seed) instead of on_construct? It would stop mod-placed seeds and plants from growing.

**Comment by sofar** [comment]
@yyt16384 

Is there any reason the mods can't use `minetest.place_node()` ?

**Comment by yyt16384** [comment]
@sofar When the mod is not placing the node, but instead moving the node around (like pistons). But I think this should be a bug for other mods, so no problems here.
